### PR TITLE
Fix in Jp2 metadata writing & improvements in reading

### LINKED ISSUE #2147: Metacopy CLI tool corrupt JPEG2000 (regression with Exiv2 >= 0.27.5)
* Context : 

Linux Mageia8
Exiv2 from 0.27-maintenance branch compiled with samples tool and installed 
One JPEG file with Exif metadata (https://drive.google.com/file/d/13n4MMb9hjUKnNuTjVcDVP7sVt0icYjTu/view?usp=sharing)
ImageMagick version 7.0.10-62 Q16 x86_64

* Steps to reproduce the bug : 

1/ I convert a JPEG containing metadata with ImageMagick:

﻿﻿[gilles@localhost JP2]$ convert 4324477432.jpg 4324477432.jp2

2/ I check if ImageMagick is able to parse JPEG2000 file.

[gilles@localhost JP2]$ identify 4324477432.jp2
4324477432.jp2 JP2 4032x3024 4032x3024+0+0 8-bit sRGB 0.000u 0:00.000

3/ ImageMagick do not copy metadata from JPEG to JPEG2000

[gilles@localhost JP2]$ exiv2 4324477432.jp2
File name       : 4324477432.jp2
File size       : 16249731 Bytes
MIME type       : image/jp2
Image size      : 4032 x 3024
4324477432.jp2: No Exif data found in the file

4/ Simple backup of JPEG2000 to compare later

[gilles@localhost JP2]$ cp 4324477432.jp2 4324477432_original.jp2

5/ I use Exiv2 metacopy CLI tool to import metadata from JPEG to JPEG2000

[gilles@localhost JP2]$ metacopy -eix 4324477432.jpg 4324477432.jp2

6/ Now ImageMagick cannot parse the JPEG2000 file:

[gilles@localhost JP2]$ identify 4324477432.jp2
identify: Stream error while reading JP2 Header box: box length is inconsistent.
 `OpenJP2' @ error/jp2.c/JP2ErrorHandler/194.
identify: unable to decode image file `4324477432.jp2' @ error/jp2.c/ReadJP2Image/347.

// ----

[gilles@localhost JP2]$ exiv2 --version
exiv2 0.27.5
[gilles@localhost JP2]$ convert --version
Version: ImageMagick 7.0.10-62 Q16 x86_64 2021-02-07 https://imagemagick.org
[gilles@localhost JP2]$ identify --version
Version: ImageMagick 7.0.10-62 Q16 x86_64 2021-02-07 https://imagemagick.org

* Important:

- I tried with other jpeg images taken with other camera, and problem is reproducible.
- With Exiv2 0.27.3, the problem is not reproducible.

### PR DESCRIPTION
**Pull request by piponazo**
This PR fix #2147 

After repeating the steps described by @cgilles I could easily reproduce the issue. Even by just creating a JP2 from scratch and then trying to read the metadata, the problem was reproducible (see unit test `Jp2Image.canWriteMetadataAndReadAfterwards`). 

The main issue was present at `Jp2Image::encodeJp2Header` when encoding the `newlen` of the Colour Specification Box. At that point we were indicating that the box was a bit bigger than it was in reality and the method `Jp2Image::readMetadata` would detect that afterwads.

I also did some improvements in `Jp2Image::readMetadata` and `Jp2Image::printStructure` so that we perform more checks as described in the JP2 standard. Due to some of these changes, I had to adapt some of the python tests. 

I will probably spend some more time in the future to make more improvements in the Jp2 support but I will do that in other branches. 

@cgilles I would appreciate if you could double check if this branch fixes your issues.

### CODE REVIEW

**Comment by cgilles** [comment]
Hi Luis,

I will compile your branch and check.

Note : i don't receive a notification from github. Strange...

Gilles

**Comment by cgilles** [comment]
Luis,

I can confirm that digiKam recompiled with Exiv2 from your branch "mainFixJp2" do not corrupt JP2000 file when exported from JPEG in Image Editor:

https://i.imgur.com/FF3iMeN.png

Best 

Gilles

**Comment by cgilles** [comment]
Luis, did your back-port your JPEG2000 fix to the 0.27-maintenance branch ?

**Comment by piponazo** [comment]
@cgilles great, thanks for double checking. I did not port yet these changes to `0.27-maintenance` but I will do it.

**Comment by cgilles** [comment]
Thanks Luis for the feedback.

The digiKam version that i use to test your branch is the future 8.0.0 where code is open to new implementations. This version is in alpha stage. I patched the Exiv2 interface in digiKam to compile with Exiv2 1.0.0 API.
The digiKam stable version 7.x is maintained in a dedicated branch and this code do not compile with Exiv2 1.0.0. It uses 0.27-maintenance API. 
If you back-port your branch to 0.27-maintenance, i will not back-port my Exiv2 Interface changes from digiKam 8 to 7. It will be safe as changes introduced in digiKam 8 are important.

Best

File: src/jp2image.cpp [hunk 1]
```diff
@@ -1,134 +1,167 @@
...
+            0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            0x01, 0x07, 0x01, 0x01, 0xff, 0x64, 0x00, 0x23, 0x00, 0x01, 0x43, 0x72, 0x65, 0x61, 0x74, 0x6f, 0x72, 0x3a,
+            0x20, 0x4a, 0x61, 0x73, 0x50, 0x65, 0x72, 0x20, 0x56, 0x65, 0x72, 0x73, 0x69, 0x6f, 0x6e, 0x20, 0x31, 0x2e,
+            0x39, 0x30, 0x30, 0x2e, 0x31, 0xff, 0x52, 0x00, 0x0c, 0x00, 0x00, 0x00, 0x01, 0x00, 0x05, 0x04, 0x04, 0x00,
+            0x01, 0xff, 0x5c, 0x00, 0x13, 0x40, 0x40, 0x48, 0x48, 0x50, 0x48, 0x48, 0x50, 0x48, 0x48, 0x50, 0x48, 0x48,
+            0x50, 0x48, 0x48, 0x50, 0xff, 0x90, 0x00, 0x0a, 0x00, 0x00, 0x00, 0x00, 0x00, 0x2d, 0x00, 0x01, 0xff, 0x5d,
+            0x00, 0x14, 0x00, 0x40, 0x40, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0xff, 0x93, 0xcf, 0xb4, 0x04, 0x00, 0x80, 0x80, 0x80, 0x80, 0x80, 0xff, 0xd9};
+
+        struct Jp2BoxHeader
```
[the remark this answers is not in the document]
Comment (piponazo) [reply src/jp2image.cpp]: Can you provide an example of what you have in mind?

**Pull request closed without merging**